pkg/storegateway/indexheader/lazy_binary_reader.go

```diff
@@ -232,27 +232,33 @@ func (r *LazyBinaryReader) load() (returnErr error) {
 		return r.readerErr
 	}
 
+	// Release the read lock, so that load1 can take write lock. Take the read lock again once done.
+	r.readerMx.RUnlock()
```

Comment:
nit: Seems like there's no real reason to acquire the read lock in the various `Reader` methods anymore and it could just be done in this method.

Reply:
This code can be simplified, but I'd suggest doing that in separate PR.

Simplification would be that `load()` returns reader and error, and then there's no need to perform locking in other `Reader` methods. Eg. `IndexVersion` would become just:

```
func (r *LazyBinaryReader) IndexVersion() (int, error) {
	reader, err := r.load() // would also update r.usedAt
	if err != nil {
		return 0, err
	}

	return r.reader.IndexVersion()
}
```

Reply:
I'll merge this, and send new PR that also renames methods.

Reply:
I tried to implement the change I described above, but fortunately there's a test that stopped me. The `r.readerMx` lock also prevents `unloadIfIdleSince` from happening while `r.reader` is still in use. That's why all the `Reader` methods need to hold read-lock.

Reply:
See here: https://github.com/grafana/mimir/pull/5608